packages/analytics-core/src/plugins/destination.ts

```diff
@@ -297,21 +313,25 @@ export class Destination implements DestinationPlugin {
   }
 
   fulfillRequest(list: Context[], code: number, message: string) {
-    this.saveEvents();
     list.forEach((context) => context.callback(buildResult(context.event, code, message)));
   }
 
   /**
-   * Saves events to storage
-   * This is called on
-   * 1) new events are added to queue; or
-   * 2) response comes back for a request
+   * update the event storage
    */
-  saveEvents() {
+  async updateEvent(dropEventSet: Set<string>) {
     if (!this.config.storageProvider) {
       return;
     }
-    const events = Array.from(this.queue.map((context) => context.event));
-    void this.config.storageProvider.set(this.storageKey, events);
+    const savedEvent = await this.config.storageProvider.get(this.storageKey);
```

Comment:
How do we handle multiple threads trying to write to the storage at the same time? What happens in a case like this
1. We get the events to be saved
2. We are filtering events
3. While we are filtering events, another thread adds a new event to storage
4. We are done filtering events and we now save the retriedEvents, which does not include the new event that was saved in 3

Reply:
Added a lock to prevent the concurrency issue. Please check.

Reply:
I guess this lock is not preventing the saving events though. Can we confirm if this is a concern in JS/TS? I recall JS is single threaded so just to confirm if we still have this race condition in JS.

Reply:
Javascript has a single thread, it executes one instruction/sequence at a time. So there is no concept of concurrent execution or parallelism.
Asynchronous means the program can work or execute another sequence of instructions while one of the sequences waits. It is responsive and lets the other instructions manage instead of making them starve. 
In this case, the code in `updateEventStorage`, between the await this.config.storageProvider.get(this.storageKey) and await this.config.storageProvider.set(this.storageKey, xxx) won't be interrupted.

Reply:
Followup. The logic for the legacy SDK is as following:
In init, we load the unsent event from the storage, and we save them in a local queue (array).
Before sending an event,  we save the event in the local queue, overwrite the storage based on the local queue. 
While there have a response, we update the local queue, and update the storage with the local queue.